File: my-solutions/0207-course-schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj_list = {i: [] for i in range(numCourses)}
        for i in range(len(prerequisites)):
            adj_list[prerequisites[i][1]].append(prerequisites[i][0])
        
        # print(adj_list)

        visited = set()
        back_edge = False

        def dfs(node):
            if node in visited:
                # print(node, "back_edge")
                nonlocal back_edge 
                back_edge = True
                return
            else:
                visited.add(node)
                # print(node, neighbours)
                for nei in adj_list[node]:
                    dfs(nei)
                visited.discard(node)
                adj_list[node] = []

        for key in adj_list:
            back_edge = False
            dfs(key)
            # print(key, back_edge)
            if back_edge:
                return False
        
        # dfs(0)
        # if back_edge:
        #         return False

        return True
        

File: my-solutions/0207-course-schedule/solution.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // dfs
        vector<int> visited(numCourses, 0);
        unordered_map<int, vector<int>> adj; // pre-req : course
        for(auto& pre: prerequisites) {
            adj[pre[1]].push_back(pre[0]);
        }

        bool res = true;

        for(int i = 0; i < numCourses; ++i) {
            res = res && dfs(i, adj, visited);
        }

        return res;
    }

    bool dfs(int course, unordered_map<int, vector<int>>& adj, vector<int>& visited) {
        visited[course] = 1;

        bool res = true;
        for(auto& neighbor: adj[course]) {
            if(visited[neighbor] == 1) return false;
            if(visited[neighbor] == 0) {
                res = res && dfs(neighbor, adj, visited);
            }
        }

        visited[course] = 2;
        return res;
    }
};
